### ISSUE TITLE
is it possible to use executeJs on frames

### ISSUE THREAD

**Issue by andynuss** [ulixee/secret-agent#321]
I have an executeJs function that does various things, none of which involve changing the dom, an example of which is to scrape dynamically created style sheets.

I wish to do this with iframes of the main window, possibly 2 levels deep, even if the iframe is a different origin than the main window.  Thus it is not possible in general to simply call executeJs() on the main window and iterate the window.frames pseudo-array to do this.

So I tried the following code, and got a runtime exception telling me that frame.use() is not a function.

    import { Agent } from 'secret-agent';
    import ExecuteJsPlugin from '@secret-agent/execute-js-plugin';

    export async function scrapeUrl(requestUrl: string): Promise<string[]> {

      const userAgent =
        'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_16_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.165 Safari/537.36';
      const agent = new Agent({
        userAgent,
      });

      await agent.goto(requestUrl);
      await agent.waitForPaintingStable();

      const frames = await agent.frameEnvironments;
      const result: string[] = [];

      // NOTE: the first frame is the main window
      for (let i = 0; i < frames.length; i++) {
        const frame = frames[i];
        // @ts-ignore
        frame.use(ExecuteJsPlugin);

        // @ts-ignore
        const html = await frames[i].executeJs(() => {
          // @ts-ignore
          const result = document.documentElement.outerHTML;
          return result;
        });
        result.push(html);
      }

      await agent.close();
      return result;
    }

Is there a way to accomplish this currently?  NOTE: if I just needed the outerHTML, I'm sure it could be done another way, but I am just trying to see if I can inject javascript into any window available from the main document's frames.

**Comment by blakebyrnes** [comment]
The ExecuteJs plug-in is really intended as an example of how to create a plug-in. There are two options that I can think of:
1) write a plug-in that accesses the underlying frames of the page that gets passed to the callback and use the evaluate function on those frames. As an added bonus, you can run the js in an isolated JavaScript context where your global variables can't be seen by the main page
2) Each frameenvironment you've accessed exposes a document that can run full awaited-dom functionality (asynchronous read only dom). Can you examine your style sheets with awaited dom features?

**Comment by andynuss** [comment]
Your second point is something to look into.  Perhaps.  Your first point is really surprising.  Does the executeJs plugin provided by you guys allow me to see how to hide my global variables from the main page?  That wasn't clear to me with the executeJs method for the main page, because it felt like it simply emulated puppeteer's page.evaluate() function.

**Comment by blakebyrnes** [comment]
When we built plugins, we exposed some of the internals of SecretAgent that don't have any documentation yet. Typescript should yield some basic type information, but it's not going to have good descriptions for what parameters mean. If you look in the CorePlugin in executeJs, you'll see a `puppetPage` that gets passed in. It comes with an "evaluate" that simply runs a `Runtime.evaluate` against the main frame of the page (https://chromedevtools.github.io/devtools-protocol/tot/Runtime/#method-evaluate).

Also on that object is an array of "frames". Those frames have a direct `evaluate` function that can be run. This is the signature:
```
evaluate<T>(
    expression: string,
    isolateFromWebPageEnvironment?: boolean,
    options?: { shouldAwaitExpression?: boolean; retriesWaitingForLoad?: number },
  ): Promise<T>
```

The second parameter is how to "isolate" from the main web page environment. As some background to "what" this is - Chromium was adapted a while back to provided isolated javascript environments to various frames based on sandbox type settings. They extended this to Chrome extension developers to be able to add variables and globals to their own environment without impacting the browser page (eg, injecting a different version of jQuery globally). 

Secret Agent keeps track of it's own "isolated" environment where all awaited dom activities and page event tracking occurs. This keeps it shielded from the main webpage and any attempts to look for the use of querySelectors and other similar calls (ie, page proxies the querySelector and waits for calls they don't recognize).

Hopefully that helps. In regards to figuring out "how" to use the different parts of "puppet", I would use Typescript (or look at the definition files), and for now, look at code snippets. We'll hopefully get some documentation up shortly that will make the rest easier.

**Comment by andynuss** [comment]
Concerning this "isolated" environment that devtools offers, am I correct in understanding this to be a way to call various window and document built-in functions without worrying about whether the webpage has already "monkey-patched" that method to intercept all its calls?

**Comment by blakebyrnes** [comment]
It's a possibility, but you'd be bypassing a large portion of what is already built-in to SecretAgent to do so. This is the built-in way: `await agent.document.documentElement.outerHTML`...

All of the abilities to replay sessions and see which elements are actually being picked up, and the ability to create reliable scripts are far less powerful when you skip out on using the built-in stuff.

Are you finding that AwaitedDom (https://secretagent.dev/docs/basic-interfaces/awaited-dom), and the built-in Document (https://secretagent.dev/docs/basic-interfaces/agent#document) aren't providing you with something you need?

**Comment by andynuss** [comment]
The thing I am doing is heuristically hiding (display none) div-based "popups" with fixed positioning, capturing the dom, plus all the dynamic sheets, capturing the screenshot, and then restoring the fixed positioned popus.  Also, I am inserting numbering attributes of my own before capturing the dom, and then removing that too.

**Comment by andynuss** [comment]
So what I am understanding (and confused about), is that there are two plugins of interest in the execute-js package, ExecuteJsCorePlugin and ExecuteJsClientPlugin, and I need to model my plugin after one of these, but which one? It must be ExecuteJsCorePlugin because that is the one that has a puppetPage involved, but I don't see how this is the one that gets selected by the following import:

    import ExecuteJsPlugin from '@secret-agent/execute-js-plugin';

... based on studying index.js in the execute-js package.  So I am a little confused about how the use() statement works.

However, assuming that I just write a plugin with the same signature ExecuteJsCorePlugin, and that my implementation of "onClientCommand" gets the puppetPage, when I call agent.use(MyPlugin), then I seems like what I am doing different than  is exposing the ability to call evaluate<T> on the frames of the puppetPage argument, and also passing true as the second argument to evaluate<T>.

But given that I am passing the class itself to agent.use(), and not an instance of the class, how do I expose my function for the main page, and my function for the evaluating on the frames?

**Comment by andynuss** [comment]
I guess what I am saying is that IClientPlugin seems to be central to the exposing of the onTab and onClient functions that allow the patching of the tab and agent with a function that ends up forwarding to the CorePlugin and its puppetPage.

But this plugin scaffolding seems to be unique to IPuppetPage and I see nothing of the sort for IPuppetFrame.

Am I missing something?

**Comment by blakebyrnes** [comment]
Sorry, I meant to write back to your first post, but I got a bit tangled reading it.

SecretAgent is built in a client/server way (Core is server, Client is ...client.. ha). You're overriding two sides: 1) enable functionality on the server (where puppetPages and frames live), and 2) add some ability to trigger that functionality in your Client (ie, client Hero/Tab/FrameEnvironment) api.

When you add a client side api, it will call through to the backend api into your `onClientCommand` callback. That callback gets a puppetPage, from which you can access the "frames" and get to the respective subframes.

What you're trying to do is definitely pretty advanced. Are you looking through the monorepo, or the node_modules code? I think you're going to need to use typescript and poke around the monorepo to pull this off.

**Comment by andynuss** [comment]
It appears that central to this plugin approach is the function argument 'sendToCore'.  I was hoping I could actually return the puppetPage in the method that sendToCore invokes.  But is it true the sendToCore can only return serializable results?

I was trying to do this:
    export class MyClientPlugin extends ClientPlugin {

      public onAgent(agent, sendToCore: ISendToCoreFn): void {
        agent.getMyPuppetPage = (fn, ...args) => {
          return this.getMyPuppetPage(fn, sendToCore, args);
        };
      }

      public onTab(tab, sendToCore: ISendToCoreFn): void {
        tab.getMyPuppetPage = (fn, ...args) => {
          return this.getMyPuppetPage(fn, sendToCore, args);
        };
      }

      private getMyPuppetPage(fn, sendToCore, args): Promise<any> {
        return sendToCore(myPluginId, '', '', args);
      }
    }

    export class MyCorePlugin extends CorePlugin {

      public async onClientCommand(
        { puppetPage }: IOnClientCommandMeta,
        fnName: string,
        serializedFn: string,
      ): Promise<any> {
        return puppetPage;
      }
    }

and was getting undefined for the result of getMyPuppetPage()

**Comment by blakebyrnes** [comment]
@andynuss Yes, this is correct. SecretAgent is built in a client/server way because it's expected that in most cases you'll run more than one server. Plugins are trying to give you a way to add client functionality as well as server

**Comment by andynuss** [comment]
I think I'm having another problem.  Clearly the plugin allows me to use sendToCore to effectively call puppetPage.evaluate() or as you said earlier, puppetFrame.evaluate().

But I was wondering if I could have multiple functions, including some that get info such as the number of iframes for the puppetPage or various stringifiable attributes of all the frames.

It seems that the call signature of sendToCore (which is the only way I see of getting access to the puppetPage on the remote side) forces me to use some kind of serializedFn that the secret-agent code understands.

This seems to explain why my code above returned an undefined result.  Do you understand what I am wanting to do?

**Comment by blakebyrnes** [comment]
I'm not sure I'm understanding fully, but you can have as many functions as you want... sendToCore takes any arbirtary arguments (as long as they're serializable). So you could pass through a second argument that's your "action" for the remote core to take: eg, `sendToCore(myPluginId, 'getFrameMetadata')` and then in Core, just route to the correct function using that parameter

```
export class MyCorePlugin extends CorePlugin {
  public async onClientCommand(
    { puppetPage }: IOnClientCommandMeta,
    action: string, ... whateverElse: any[],
  ): Promise<any> {
   if (action === 'getFrameMetadata') return this.getFrameMetdata();
    return puppetPage;
  }

  private getFrameMetadata() {
   ....
 }
}
```

Regarding the "serializedFn" -- that's literally a function that will be called in the browser page. It doesn't have any SecretAgent knowledge.

**Comment by andynuss** [comment]
Maybe you could show me how I would write a simple function that takes a variable number of args, and returns an array consisting of those args, using the plugin architecture and sendToCore?  That would help me see what I am missing.  I did see how the "action" arg could be used to call multiple functions.

**Comment by andynuss** [comment]
i.e. a simple echoing function

**Comment by blakebyrnes** [comment]
Is this helpful?

```
  private echo(echo1: string, echo2: number, ...args:any[]): Promise<any> {
    const echoed = await sendToCore(myPluginId, echo1, echo2 ...args);
    console.log(echoed); // echo1 string, echo2 number, ... rest
    return echoed;
  }
}

export class MyCorePlugin extends CorePlugin {
  public async onClientCommand(
    { puppetPage }: IOnClientCommandMeta,
   echo1: string,
   echo2: number,
   ...args: any[]
  ): Promise<any> {
    return [echo1, echo2, ...args]
  }
}
````

**Comment by andynuss** [comment]
Yes! I hope I can get it to work, or else I am doing something really more stupid, such as with myPluginId

**Comment by andynuss** [comment]
Still having some problem, probably not with the part you gave me:


    import ClientPlugin from '@secret-agent/plugin-utils/lib/ClientPlugin';
    import CorePlugin from '@secret-agent/plugin-utils/lib/CorePlugin';
    import { ISendToCoreFn } from '@secret-agent/interfaces/IClientPlugin';
    import { IOnClientCommandMeta } from '@secret-agent/interfaces/ICorePlugin';

    const myPluginId = 'my-test-plugin';

    export class MyClientPlugin extends ClientPlugin {

      public onAgent(agent, sendToCore: ISendToCoreFn): void {
        agent.echo = (echo1: string, echo2: number, ...args:any[]) => {
          return this.echo(sendToCore, echo1, echo2, args);
        };
      }

      public onTab(tab, sendToCore: ISendToCoreFn): void {
        tab.echo = (echo1: string, echo2: number, ...args:any[]) => {
          return this.echo(sendToCore, echo1, echo2, args);
        };
      }

      private async echo(sendToCore: ISendToCoreFn, echo1: string, echo2: number, ...args:any[]): Promise<any> {
        const echoed = await sendToCore(myPluginId, echo1, echo2, ...args);
        console.log('sendToCore result', echoed);
        return echoed;
      }
    }

    export class MyCorePlugin extends CorePlugin {

      public async onClientCommand(
        { puppetPage }: IOnClientCommandMeta,
        echo1: string,
        echo2: number,
        ...args: any[]
      ): Promise<any> {
        return [echo1, echo2, ...args];
      }
    }


... but with the test function that uses it after scraping a page:

    import { Agent, ConnectionFactory, ConnectionToCore, LocationStatus } from 'secret-agent';
    import { MyClientPlugin, MyCorePlugin } from './plugin';

    const myplugins = { MyClientPlugin, MyCorePlugin };

    const ua = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_16_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.165 Safari/537.36';

    let sharedConnection: ConnectionToCore = null;

    function getConnection (): ConnectionToCore {
      if (sharedConnection !== null) return sharedConnection;
      sharedConnection = ConnectionFactory.createConnection({
        maxConcurrency: 4,
      });
      return sharedConnection;
    }

    export async function testUrl(requestUrl: string, imageName: string): Promise<void> {
      const myagent = new Agent({
        userAgent: ua,
        viewport: {
          screenHeight: 768,
          screenWidth: 1024,
          height: 768,
          width: 1024,
        },
        connectionToCore: getConnection(),
      });

      try {
        myagent.use(myplugins);
        await myagent.goto(requestUrl);
        await myagent.waitForPaintingStable();
        await myagent.activeTab.waitForLoad(LocationStatus.DomContentLoaded);

        // @ts-ignore
        const echoargs = await myagent.echo('a', 1, 'b', 2);
        console.log('echoargs', echoargs);

      } finally {
        try {
          await myagent.close();
        } catch (e) {
          console.log('unexpected error closing new Agent', e);
        }
      }
    }

these are my log lines:

    sendToCore result undefined
    echoargs undefined

**Comment by blakebyrnes** [comment]
I think it might be because you're not actually assigning the ID to the plugins. So it might not know how to route it. @calebjclark, is that right?

**Comment by andynuss** [comment]
NOTE: I was never really sure how to set up the pluginId.  My code above that is not working was my simple attempt, so please let me know what to do.

**Comment by blakebyrnes** [comment]
Try adding `static readonly id = 'my-test-plugin'` into Core/Client plugins

**Comment by andynuss** [comment]
How would that help?  In my main test function, I do this:

    import { MyClientPlugin, MyCorePlugin } from './plugin';

    const myplugins = { MyClientPlugin, MyCorePlugin };
    ....
    myagent.use(myplugins);

and I don't think a static readonly id would be part of the 'myplugins' object that I pass to the use() function.

**Comment by andynuss** [comment]
Oh, I see what you meant.  One sec

**Comment by calebjclark** [comment]
@andynuss, it's required for SecretAgent to hook into the plugin. It's how we designed the plugin architecture.

**Comment by andynuss** [comment]
I think I did what you asked, and still getting the same undefined result as before.  Here is my compiled javascript for plugin.js:

    "use strict";
    var __importDefault = (this && this.__importDefault) || function (mod) {
        return (mod && mod.__esModule) ? mod : { "default": mod };
    };
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.MyCorePlugin = exports.MyClientPlugin = void 0;
    const ClientPlugin_1 = __importDefault(require("@secret-agent/plugin-utils/lib/ClientPlugin"));
    const CorePlugin_1 = __importDefault(require("@secret-agent/plugin-utils/lib/CorePlugin"));
    class MyClientPlugin extends ClientPlugin_1.default {
        onAgent(agent, sendToCore) {
            agent.echo = (echo1, echo2, ...args) => {
                return this.echo(sendToCore, echo1, echo2, args);
            };
        }
        onTab(tab, sendToCore) {
            tab.echo = (echo1, echo2, ...args) => {
                return this.echo(sendToCore, echo1, echo2, args);
            };
        }
        async echo(sendToCore, echo1, echo2, ...args) {
            const echoed = await sendToCore(MyClientPlugin.id, echo1, echo2, ...args);
            console.log('sendToCore result', echoed);
            return echoed;
        }
    }
    exports.MyClientPlugin = MyClientPlugin;
    MyClientPlugin.id = 'my-test-plugin';
    class MyCorePlugin extends CorePlugin_1.default {
        async onClientCommand({ puppetPage }, echo1, echo2, ...args) {
            return [echo1, echo2, ...args];
        }
    }
    exports.MyCorePlugin = MyCorePlugin;
    MyCorePlugin.id = 'my-test-plugin';

**Comment by andynuss** [comment]
(I think I've given you all the pieces of the code that's not working for me.)

**Comment by calebjclark** [comment]
@andynuss, your undefined problem is being caused by two issues:

1) For security reasons, ClientPlugins must explicitly state what core plugins they depend on. Add a `static coreDependencyIds` array to your MyClientPlugin:

```
export class MyClientPlugin extends ClientPlugin {
  static id = 'my-test-plugin';
  static coreDependencyIds = ['my-test-plugin'];
  ...
}
```

2) Your MyCorePlugin cannot be passed as an object into agent.use. The reason is that the Client and Core sides of SecretAgent are actually running in separate NodeJs processes. The easiest approach in your situation is to just pass an absolute path. Change from this:
```
import { MyClientPlugin, MyCorePlugin } from './plugin';

const myplugins = { MyClientPlugin, MyCorePlugin };
....
myagent.use(myplugins);
```
to this:
```
import Path from 'path';
....
myagent.use(Path.resolve(__dirname, 'plugin'));

```
Let me know if that works.

Long story short, this entire flow is really confusing -- we're fixing this in our next big release.

**Comment by andynuss** [comment]
I tried both parts of the solution, and it seems like this call:

    myagent.use(Path.resolve(__dirname, 'plugin'));

... which did succeed, caused the following line:

    await myagent.goto(requestUrl);

with this error:

    testUrl failed: TypeError: coreTab.goto is not a function
    at Tab.goto (/Users/andynuss/repos/stag-secret-agent/app/node_modules/client/lib/Tab.ts:161:36)

**Comment by calebjclark** [comment]
@andynuss, it's tough to know what's going on without seeing the full client script as well. Any chance you can send me your full script(s) -- both plugins + your client script?

Either temporarily add me to a github repo as a dev or email it to me (caleb@ulixee.org). I'm glad to investigate further.

**Comment by andynuss** [comment]
will do

**Comment by andynuss** [comment]
do you mind if I send the compiled js files, or would you rather have the 2 .ts files?

**Comment by calebjclark** [comment]
@andynuss please send them as Ts files.

**Comment by calebjclark** [comment]
@andynuss what version of secretagent is in your package.json?

**Comment by andynuss** [comment]
"dependencies": {
      "ts-node": "^10.1.0",
      "secret-agent": "^1.5.5",
      "@secret-agent/execute-js-plugin": "^1.5.5",
      "@secret-agent/plugin-utils": "^1.5.5",
      "@secret-agent/interfaces": "^1.5.5"
    },

**Comment by blakebyrnes** [comment]
Sorry for the delay. Think we figured out the part of the .goto blowing up

**Comment by andynuss** [comment]
So, should I try it again with 1.5.10?

**Comment by blakebyrnes** [comment]
@andynuss yes, you should get past the previous client errors.

**Comment by andynuss** [comment]
I am still getting the same problem last identified with the 'coreTab.goto is not a function'

using the same two files I gave you by email before -- plugin.ts and test.ts

what does this do?  Is it possible that my directory structure is different?
    myagent.use(Path.resolve(__dirname, 'plugin'));

also, I am running my test.js/plugin/js pair in a different directory than where I have my plugin.ts and test.ts files located.
and, this is a biggie, I am invoking the main function in test.js from my own simple nodejs http service (server.js) that is in turn triggered by a remote (java) program sending an url to the service

But to test out my theory that it had something to do with using test.js from my server.js http listener, I added this to the
bottom of test.ts:

    (async () => {
      await testUrl('https://www.whatsmyua.info', 'dummy');
    })();

... and sure enough, doing 'node test.js' caused the same gotoTab error

---

This indicates to me that either my compiled test.js/plugin.js files are different in some critical way than yours,
or there is some kind of relative directory problem in the call to this:
    myagent.use(Path.resolve(__dirname, 'plugin'));

**Comment by blakebyrnes** [comment]
I think you sent your files to Caleb. I haven't seen them. Your setup sounds like you might want to start your own Core inside your server (https://secretagent.dev/docs/advanced/remote). I think you're having issues with Core spawning off into a new process. That said, it's very difficult to tell more about what's going on without more code or logs

**Comment by andynuss** [comment]
The only thing I know to do is send my pair of *.ts files as well as my pair of *.js files and ask you to simply run 'test.js' with node.  Do you want me to send these files somewhere?

**Comment by blakebyrnes** [comment]
Sure - send them to me at blake at ulixee.org. If you have a session database for a failed run, that might help too - https://secretagent.dev/docs/advanced/session

**Comment by andynuss** [comment]
I see that a fix has been merged.  Should my code "just work" if I use the latest version?

**Comment by blakebyrnes** [comment]
Take a look at the examples I checked in. Your code might need some tweaks: https://github.com/ulixee/secret-agent/blob/main/examples/plugin-echo-use.ts

**Comment by andynuss** [comment]
I was doing this:

    import ClientPlugin from '@secret-agent/plugin-utils/lib/ClientPlugin';
    import CorePlugin from '@secret-agent/plugin-utils/lib/CorePlugin';

and your example does this:

    import { ClientPlugin, CorePlugin } from '@secret-agent/plugin-utils';

I take it that these are not equivalent and I need to do it your way?

**Comment by blakebyrnes** [comment]
Both options are doing the same thing. I think we had forgotten to export from top level when you created your first example, but it was listed in the docs.

**Comment by andynuss** [comment]
Well, does it surprise you that I simply did a 'yarn install' for 1.5.11 and my example (the one given to you guys) just worked untouched?

**Comment by andynuss** [comment]
Thanks!
